### PR TITLE
Fix parentheses around return type annotations

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

This PR addresses the issue where subscript and other parenthesized expressions in function return type positions were wrapped in an extra pair of parentheses if the function has no parameters. This leads to rather verbose code: 

```patch
 @pytest.fixture
-def mock_legacy_device_tracker_setup() -> (
-    Callable[[HomeAssistant, "MockScanner"], None]
-):
+def mock_legacy_device_tracker_setup() -> Callable[
+    [HomeAssistant, "MockScanner"], None
+]:
     """Return setup callable for legacy device tracker setup."""
     from tests.components.device_tracker.common import mock_legacy_device_tracker_setup

```

It's also inconsistent to how the same return type gets formatted when the function has parameters:

```python
@pytest.fixture
def mock_legacy_device_tracker_setup(
    a,
) -> Callable[
    [HomeAssistant, "MoooooooooooooooooooooooooooooooooooooooooooockScanner"], None
]:
    """Return setup callable for legacy device tracker setup."""
    from tests.components.device_tracker.common import mock_legacy_device_tracker_setup
```

Fixes https://github.com/astral-sh/ruff/issues/9447

## Black deviation

The PR adds two fairly extensive tests that walk through the black deviations. The most notable deviation is that black tries to parenthesize types before splitting the type itself (only applies to subscript)

**Black**
```python
def func() -> (
    list[1, 2222222222222222222, 3333333333333333333333333333333333333333333333333]
):
    pass
```

But it only does so if the function has **no** parameters. Below is Black's formatting if the function has a parameter.

```python

def func(
    a,
) -> list[
    1, 2222222222222222222, 33333333333333333333333333333333333333333333333333333
]:
    pass


def func(
    a,
) -> list[1, 2222222222222222222, 3333333333333333333333333333333333333333333333333333]:
    pass
```

This seems inconsistent. Ruff only parenthesizes the `list` if the subscript value up to the `[` exceeds the line length

**Ruff**

```python
def func() -> list[
    1, 2222222222222222222, 3333333333333333333333333333333333333333333333333
]:
    pass


def func(
    a,
) -> list[
    1, 2222222222222222222, 33333333333333333333333333333333333333333333333333333
]:
    pass


def func(
    a,
) -> list[1, 2222222222222222222, 3333333333333333333333333333333333333333333333333333]:
    pass


# Ruff uses parentheses here similar to Black
def thiiiiiiiiiiiiiiiiiis_iiiiiiiiiiiiiiiiiiiiiiiiiiiiiis_veeeeeeeeeeedooooong() -> (
    list[int, float]
): ...

```

The main counterargument is that most changes in the ecosystem are cases where Black would parenthesize the subscript. I'm interested to get a second opinion on this.

> **Note**: Supporting black's formatting would introduce a fair amount of extra complexity. We couldn't use `maybe_parenthesize_expression` anymore and need to use `best_fitting` to model the three cases (type fits, type parenthesized fits, split the type)

## Preview

This is a preview-only change because it changes the formatting of already formatted code.

## Test Plan

I reformatted some projects locally and skimmed through the changes. Overall, there are only few, and I generally prefer the new formatting (see [this comment](https://github.com/astral-sh/ruff/pull/13381#issuecomment-2355998843)). Whether one prefers Black's formatting with the extra parentheses or not is probably the only "questionable" formatting change. IMO, consistency is more important (it reduces surprises for users: Why is it parenthesized here?).

### CODE REVIEW

**Comment by MichaReiser** [comment]
Some more examples:

```patch
 @cache
-def _register_filesystems() -> (
-    Mapping[
-        str,
-        Callable[[str | None, Properties], AbstractFileSystem] | Callable[[str | None], AbstractFileSystem],
-    ]
-):
+def _register_filesystems() -> Mapping[
+    str,
+    Callable[[str | None, Properties], AbstractFileSystem] | Callable[[str | None], AbstractFileSystem],
+]:
     scheme_to_fs = _BUILTIN_SCHEME_TO_FS.copy()
     with Stats.timer("airflow.io.load_filesystems") as timer:

```

File: crates/ruff_python_formatter/resources/test/fixtures/ruff/expression/hug.py [hunk 1]
```diff
@@ -82,30 +82,6 @@
 
 func([(x, y,) for (x, y) in z], bar)
 
-# Ensure that return type annotations (which use `parenthesize_if_expands`) are also hugged.
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/resources/test/fixtures/ruff/expression/hug.py:85]: This is no longer true. We never add parentheses around lists anymore

**Comment by MichaReiser** [comment]
TODO: Ruff, for some reason, preserves the parentheses around:

```python
def thiiiiiiiiiiiiiiiiiis_iiiiiiiiiiiiiiiiiiiiiiiiiiiiiis_veeeeeeeeeeedooooong() -> (
    list[int, float]
): ...
```

File: crates/ruff_python_formatter/tests/snapshots/format@statement__return_annotation.py.snap [hunk 2]
```diff
@@ -521,4 +513,22 @@ def process_board_action(
...
+@@ -249,11 +249,8 @@
+ ): ...
+ 
+ 
+-def double() -> (
+-    first_item
+-    and foo.bar.baz().bop(
+-        1,
+-    )
++def double() -> first_item and foo.bar.baz().bop(
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/tests/snapshots/format@statement__return_annotation.py.snap:529]: This looks surprising but it matches black's formatting

**Comment by MichaReiser** [comment]
Whoops, I must have missed a preview check somewhere.

**Review by charliermarsh — APPROVED** [review]
Interesting... I do prefer what you implemented here. I was looking back at https://github.com/astral-sh/ruff/pull/6436 -- I'm wondering why we didn't make this change then? I guess we were biasing towards improving compatibility as much as possible, despite us not understanding why Black had different styling at that point in time.

**Comment by MichaReiser** [comment]
> Interesting... I do prefer what you implemented here. I was looking back at #6436 -- I'm wondering why we didn't make this change then? I guess we were biasing towards improving compatibility as much as possible, despite us not understanding why Black had different styling at that point in time.

Looking through the tests added in #6436 the once that are relevant for this change are

```diff
-def xxxxxxxxxxxxxxxxxxxxxxxxxxxx() -> (
-    Set[
-        "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
-    ]
-): ...
+def xxxxxxxxxxxxxxxxxxxxxxxxxxxx() -> Set[
+    "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
+]: ...
 
 
-def xxxxxxxxxxxxxxxxxxxxxxxxxxxx() -> (
-    Set[
-        "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
-    ]
-): ...
+def xxxxxxxxxxxxxxxxxxxxxxxxxxxx() -> Set[
+    "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
+]: ...
 
 
-def xxxxxxxxxxxxxxxxxxxxxxxxxxxx() -> (
-    Set[
-        "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
-    ]
-): ...
+def xxxxxxxxxxxxxxxxxxxxxxxxxxxx() -> Set[
+    "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
+]: ...
 
 
-def xxxxxxxxxxxxxxxxxxxxxxxxxxxx() -> (
-    Set[
-        "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
-    ]
-): ...
+def xxxxxxxxxxxxxxxxxxxxxxxxxxxx() -> Set[
+    "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
+]: ...
```

These are all cases where black keeps the parentheses. But it removes them as soon as the subscript contains another element

```python
def xxxxxxxxxxxxxxxxxxxxxxxxxxxx() -> Set[
    "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
    "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
]: ...
 ```

That's why I see this mainly as a refinement of the formatting you implemented in #6436.

**Pull request merged**